addons/html_editor/static/src/core/dom_plugin.js

```diff
@@ -147,37 +154,34 @@ export class DomPlugin extends Plugin {
         }
 
         startNode = startNode || this.shared.getEditableSelection().anchorNode;
-        // If the selection anchorNode is the editable itself, the content
-        // should not be unwrapped.
-        if (selection.anchorNode !== this.editable) {
-            // In case the html inserted is all contained in a single root <p> or <li>
-            // tag, we take the all content of the <p> or <li> and avoid inserting the
-            // <p> or <li>. The same is true for a <pre> inside a <pre>.
-            if (
-                container.childElementCount === 1 &&
-                (container.firstChild.nodeName === "P" ||
-                    container.firstChild.nodeName === "LI" ||
-                    (container.firstChild.nodeName === "PRE" && closestElement(startNode, "pre")))
-            ) {
-                const p = container.firstElementChild;
-                container.replaceChildren(...p.childNodes);
-            } else if (container.childElementCount > 1) {
-                // Grab the content of the first child block and isolate it.
-                if (
-                    isBlock(container.firstChild) &&
-                    !["TABLE", "UL", "OL"].includes(container.firstChild.nodeName)
-                ) {
-                    containerFirstChild.replaceChildren(...container.firstElementChild.childNodes);
-                    container.firstElementChild.remove();
-                }
-                // Grab the content of the last child block and isolate it.
-                if (
-                    isBlock(container.lastChild) &&
-                    !["TABLE", "UL", "OL"].includes(container.lastChild.nodeName)
-                ) {
-                    containerLastChild.replaceChildren(...container.lastElementChild.childNodes);
-                    container.lastElementChild.remove();
-                }
+        const shouldUnwrap = (node) =>
+            [...paragraphRelatedElements, "LI"].includes(node.nodeName) &&
+            block.textContent !== "" &&
+            (block.nodeName === node.nodeName ||
+                ["BLOCKQUOTE", "PRE", "DIV"].includes(block.nodeName)) &&
+            selection.anchorNode !== this.editable;
+
+        // In case the html inserted is all contained in a single root <p> or <li>
+        // tag, we take the all content of the <p> or <li> and avoid inserting the
+        // <p> or <li>.
+        if (
+            container.childElementCount === 1 &&
+            (["P", "LI"].includes(container.firstChild.nodeName) ||
+                shouldUnwrap(container.firstChild)) &&
+            selection.anchorNode !== this.editable
+        ) {
+            const p = container.firstElementChild;
+            container.replaceChildren(...p.childNodes);
+        } else if (container.childElementCount > 1) {
+            // Grab the content of the first child block and isolate it.
```

Comment:
I would move the declarations of `isSelectionAtStart` and `isSelectionAtEnd` in the `if` here because this is the only place where they are used and they are defined 50 lines above. Or is it because the selection changes between 50 lines above and now and you need the original information?

Reply:
> Or is it because the selection changes between 50 lines above and now and you need the original information?

Previously the `splitTextNode` at start of insert would change the selection offset. But now in phoenix since we use `getEditableSelection`, it no longer changes the selection offset so we can move it above this `if`. :+1: I'll update it.